explorer/src/Pos/Explorer/Web/Server.hs

```diff
@@ -537,74 +539,72 @@ getTxSummary cTxId = do
             , ctsOutputs         = map (second mkCCoin) txOutputs
             }
 
-getRedeemAddressCoinPairs :: ExplorerMode ctx m => m [(Address, Coin)]
-getRedeemAddressCoinPairs = do
-
-    let addressCoinPairs :: [(Address, Coin)]
-        addressCoinPairs = utxoToAddressCoinPairs (unGenesisUtxo genesisUtxo)
-
-        redeemOnly :: [(Address, Coin)]
-        redeemOnly = filter (isRedeemAddress . fst) addressCoinPairs
-
-    pure redeemOnly
+data GenesisSummaryInternal = GenesisSummaryInternal
+    { gsiNumRedeemed            :: !Int
+    , gsiRedeemedAmountTotal    :: !Coin
+    , gsiNonRedeemedAmountTotal :: !Coin
+    }
 
 getGenesisSummary
     :: ExplorerMode ctx m
     => m CGenesisSummary
 getGenesisSummary = do
-    redeemAddressCoinPairs <- getRedeemAddressCoinPairs
-    redeemAddressInfo <- mapM (uncurry getRedeemAddressInfo) redeemAddressCoinPairs
-    let (cgsNumRedeemed, redeemedAmountTotal, nonRedeemedAmountTotal) =
-            foldr folder (0, mkCoin 0, mkCoin 0) redeemAddressInfo
-    let cgsNumTotal = length redeemAddressCoinPairs
+    grai <- getGenesisRedeemAddressInfo
+    redeemAddressInfo <- V.mapM (uncurry getRedeemAddressInfo) grai
+    let GenesisSummaryInternal {..} =
+            V.foldr folder (GenesisSummaryInternal 0 (mkCoin 0) (mkCoin 0))
+            redeemAddressInfo
+    let numTotal = length grai
     pure CGenesisSummary
-        { cgsNumNotRedeemed = cgsNumTotal - cgsNumRedeemed
-        , cgsRedeemedAmountTotal = mkCCoin redeemedAmountTotal
-        , cgsNonRedeemedAmountTotal = mkCCoin nonRedeemedAmountTotal
-        , ..
+        { cgsNumTotal = numTotal
+        , cgsNumRedeemed = gsiNumRedeemed
+        , cgsNumNotRedeemed = numTotal - gsiNumRedeemed
+        , cgsRedeemedAmountTotal = mkCCoin gsiRedeemedAmountTotal
+        , cgsNonRedeemedAmountTotal = mkCCoin gsiNonRedeemedAmountTotal
         }
   where
     getRedeemAddressInfo
         :: (MonadDBRead m, MonadThrow m)
-        => Address -> Coin -> m (Bool, Coin, Coin)
+        => Address -> Coin -> m GenesisSummaryInternal
     getRedeemAddressInfo address initialBalance = do
         currentBalance <- fromMaybe (mkCoin 0) <$> EX.getAddrBalance address
         if currentBalance > initialBalance then
             throwM $ Internal $ sformat
                 ("Redeem address "%build%" had "%build%" at genesis, but now has "%build)
                 address initialBalance currentBalance
         else
-            let isRedeemed = currentBalance == mkCoin 0
+            let isRedeemed = if currentBalance == mkCoin 0 then 1 else 0
```

Comment:
zeroCoins = mkCoin 0? It's repeating all over the place.

Reply:
Replaced with `minBound`.